File: limma/domain/src/main/r/limma.R
library(limma)  # imported by shim()

## NOTE
## Performs limma on a data matrix 
## Input two files: 1) data matrix; 2) Sample definition matrix
## Sample definition matrix contains sample number and status (tab delim)
## Status 1 = Experiment
## Status 0 = Control
## Status -1= Not included in analysis
##
## Data Matrix Assumption
## 
## 1) No "\'" in the input data file
## 2) All NaN is converted in NA
##
## DATE: 9.30.2013
## Author: Yaoyu
##
##
## NOTE
## Add in variable BETA as the threshold for Log Fold Change
##
## DATE: 01.13.2013
## Author: Yaoyu
##
## NOTE
## Remove SIGGENE_OUT file definition.  All filtering is to be done at the user level
## DATE 01.14.2014
## Author: Yaoyu
##
## Note
## Adjust input matrix to handle negative values in matrix
## Add in readjustment for Average Expression column 


## definte limma function
run.limma<-function(in.mtx, Experiment=NA, Control=NA){
    
  CurrMtx=in.mtx
  CurrMtx=in.mtx[,Experiment]
  
  CurrMtx=cbind(CurrMtx, in.mtx[,Control])
  
  Tissue=array(NA, dim=c(1,length(colnames(CurrMtx))))
  Tissue[1:length(Experiment)]="Experiment"
  Tissue[(length(Experiment)+1):length(colnames(CurrMtx))]="Control"
  
  Tissue=factor(Tissue,levels=c("Experiment","Control"))
  design=model.matrix(~0+Tissue)
  colnames(design)=c("Experiment", "Control")
  
  fit<-lmFit(CurrMtx, design)
  efit<-eBayes(fit)
  
  contrast.matrix=makeContrasts(ExpvsContr=Experiment-Control, levels=design)
  fit<-eBayes(contrasts.fit(fit, contrast.matrix[,"ExpvsContr"]))
  result_full=topTable(fit, adjust="fdr", number=dim(CurrMtx)[1])
  return(result_full)
}   


##
##
## read files
in.mtx=read.table(INFILE, header=TRUE, sep="\t", nrow=-1)

##
## Assume the file starts with \t and the first column is gene id
rownames(in.mtx)=in.mtx[,1]
in.mtx=na.omit(in.mtx[,-1])
in.mtx=data.matrix(in.mtx)

## Check to determine if matrix contains negative values
## off set the matrix to operate in positive values for limma
## starting with 0
min.val=min(in.mtx)
in.mtx=if(min.val<0){in.mtx+min.val*-1}else{in.mtx}


## Assign group
sample.mtx=read.table(SAMPLE_FILE, header=FALSE, sep="\t")
EXP=sample.mtx[(sample.mtx[,2]==1),1]
CON=sample.mtx[(sample.mtx[2]==0),1]

##
##  Run limma
result=run.limma(in.mtx,
  Experiment=EXP,
  Control=CON)


## Condition to add the ID column for limma version AFTER R 3.2.1, which return 6 column without
## the ID column .  Updated 09/28/2015
if(dim(result)[2]==6){
	ID=rownames(result)
	result=cbind(ID, result)
}



##
## reassign colnames
colnames(result)=c("ID", "Log Fold Change", "Average Expression", "t", "P-value", "q-value", "B")

## If the matrix contains negative values,  
## adjust the offset back to its original input values
##
result[,"Average Expression"]=if(min.val<0){result[,"Average Expression"]-(min.val*-1)}else{result[,"Average Expression"]}

write.table(result[,-c(4,7)], 
            file=RESULT_OUT, quote=FALSE, sep="\t", row.names=FALSE, col.names=FALSE)

File: src/main/java/edu/dfci/cccb/mev/analysis/r/limma/Limma.java
/*
 * The MIT License (MIT)
 * Copyright (c) 2017 Dana-Farber Cancer Institute
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the
 * Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */
package edu.dfci.cccb.mev.analysis.r.limma;

import static javax.persistence.CascadeType.ALL;

import java.util.List;
import java.util.Map;

import javax.persistence.ElementCollection;
import javax.persistence.Entity;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;
import javax.ws.rs.GET;
import javax.ws.rs.PUT;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonProperty;

import edu.dfci.cccb.mev.analysis.Define;
import edu.dfci.cccb.mev.analysis.Resolve;
import edu.dfci.cccb.mev.analysis.r.Adapter;
import edu.dfci.cccb.mev.analysis.r.R;
import edu.dfci.cccb.mev.dataset.Dataset;
import io.fabric8.annotations.Path;

/**
 * Limma analysis
 * 
 * @author levk
 */
@Entity
@R ("library(limma);\n" + "dataset <- as.data.frame (dataset);\n"
    + "run.limma<-function(in.mtx, Experiment=NA, Control=NA){\n" + "  CurrMtx=in.mtx;\n"
    + "  CurrMtx=in.mtx[,Experiment];\n" + "  CurrMtx=cbind(CurrMtx, in.mtx[,Control]);\n"
    + "  Tissue=array(NA, dim=c(1,length(colnames(CurrMtx))));\n" + "  Tissue[1:length(Experiment)]='Experiment';\n"
    + "  Tissue[(length(Experiment)+1):length(colnames(CurrMtx))]='Control';\n"
    + "  Tissue=factor(Tissue,levels=c('Experiment','Control'));\n" + "  design=model.matrix(~0+Tissue);\n"
    + "  colnames(design)=c('Experiment', 'Control');\n" + "  fit<-lmFit(CurrMtx, design);\n" + "  efit<-eBayes(fit);\n"
    + "  contrast.matrix=makeContrasts(ExpvsContr=Experiment-Control, levels=design);\n"
    + "  fit<-eBayes(contrasts.fit(fit, contrast.matrix[,'ExpvsContr']));\n"
    + "  result_full=topTable(fit, adjust='fdr', number=dim(CurrMtx)[1]);\n" + "  return(result_full);\n" + "};\n"
    + "in.mtx=dataset;\n"

    // Check to determine if matrix contains negative values
    // off set the matrix to operate in positive values for limma
    // starting with 0
    + "min.val=min(in.mtx,na.rm=TRUE);\n" + "in.mtx=if(min.val<0){in.mtx+min.val*-1}else{in.mtx};\n"
    // Assign group
    + "EXP=experiment;\n" + "CON=control;\n"
    // Run limma
    + "result=run.limma(in.mtx, Experiment=EXP, Control=CON);\n"
    // Condition to add the ID column for limma version AFTER R 3.2.1, which
    // return 6 column without the ID column . Updated 09/28/2015
    + "if(dim(result)[2]==6){ ID=rownames(result); result=cbind(ID, result); };\n"

    // reassign colnames
    + "colnames(result)=c('ID', 'Log Fold Change', 'Average Expression', 't', 'P-value', 'q-value', 'B');\n"

    // If the matrix contains negative values,
    // adjust the offset back to its original input values
    + "result[,'Average Expression']=if(min.val<0){result[,'Average Expression']-(min.val*-1)}else\n"
    + "{result[,'Average Expression']};\n"

    + "unname(apply(result,1,function(x)"
    + "  list(id=unname(x[1]),logFc=as.numeric(unname(x[2])),averageExpression=as.numeric(unname(x[3])),"
    + "       pValue=as.numeric(unname(x[5])),t=as.numeric(unname(x[4])),qValue=as.numeric(unname(x[6])))))")
public class Limma extends Adapter {
  /**
   * Dataset
   */
  private @Define @OneToOne (cascade = ALL) Dataset dataset;
  /**
   * Experiment key set
   */
  private @Define @ElementCollection List <String> experiment;
  /**
   * Control key set
   */
  private @Define @ElementCollection List <String> control;
  /**
   * Result set
   */
  private @Resolve @OneToMany (cascade = ALL) Map <String, LimmaEntry> result;

  /**
   * @param dataset
   */
  @PUT
  @Path ("dataset")
  @JsonProperty (required = false)
  public void dataset (Dataset dataset) {
    this.dataset = dataset;
  }

  /**
   * @param experiment
   */
  @PUT
  @Path ("experiment")
  @JsonProperty (required = false)
  public void experiment (List <String> experiment) {
    this.experiment = experiment;
  }

  /**
   * @param control
   */
  @PUT
  @Path ("control")
  @JsonProperty (required = false)
  public void control (List <String> control) {
    this.control = control;
  }

  /**
   * @return result
   */
  @GET
  @Path ("result")
  @JsonIgnore
  public Map <String, LimmaEntry> result () {
    return result;
  }
}
